Parse subscription details into persisted metadata

JorgeMucientes opened the pull request:
Part of: #10124 
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
Parses `SubscriptionDetails` values into `metadata` field so we can then sync any updated values for subscription products with the API

### Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->
1. Open an existing subscription product
2. Edit either the price or the expiration date
3. Save the product
4. Check that the values have been correctly updated, either by checking the product in web wp-admin or refreshing the product list and checking the updated values are still there. 

 🔂  Repeat the same steps for a product variation and verify the variation values are synced

JorgeMucientes committed:
Parse subscription details into persisted metadata

JorgeMucientes committed:
Minor code refactor for better readability

JorgeMucientes committed:
Update FluxC changeset to sync metadata values

JorgeMucientes committed:
Use extension function style

JorgeMucientes committed:
Fix bug when setting subscription length

JorgeMucientes committed:
Fix comment

JorgeMucientes committed:
Fix: show save button when variation values change

JorgeMucientes committed:
Support syncing subscription details for product variations

JorgeMucientes committed:
Update fluxc changeset for testing

JorgeMucientes commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/variations/VariationDetailViewModel.kt:
```diff
@@ -272,7 +272,11 @@ class VariationDetailViewModel @Inject constructor(
                 length = updatedLength
             )
-            viewState = viewState.copy(variation = variation.copy(subscriptionDetails = updatedSubscription))
+            val updatedVariation = variation.copy(subscriptionDetails = updatedSubscription)
+            viewState = viewState.copy(
+                variation = updatedVariation,
+                isDoneButtonVisible = updatedVariation != variation
+            )
```
This is unrelated to the main changes of the PR. Fixes a small issue were we'd change the product variation values but we won't show the "Save" button upon editing any of the subscription details.

JorgeMucientes commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/variations/VariationDetailViewModel.kt:
```diff
@@ -445,8 +449,7 @@ class VariationDetailViewModel @Inject constructor(
     }
 
     fun onSubscriptionExpirationChanged(selectedExpirationValue: Int) {
-        val newLength = if (selectedExpirationValue == 0) -1 else selectedExpirationValue
```
This is a leftover and not needed anymore. Default value for subscription `length` is now null

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/model/Product.kt:
```diff
@@ -407,6 +409,31 @@ fun Product.toDataModel(storedProductModel: WCProductModel? = null): WCProductMo
 
+    // Add any other editable product metadata here.
+    // Currently, only subscription details metadata is editable from the app. The rest is read only
+    fun SubscriptionDetails.toMetadataJson(): String {
```
WDYT about moving the function to `SubscriptionDetails` itself to avoid duplication?

We can then have something like this:

```kotlin
Product(...) {
     // Add any other editable product metadata here.
    // Currently, only subscription details metadata is editable from the app. The rest is read only
    private fun metadataJson() = subscription.toMetadataJson()
    ....
    
    it.metadata = metadataJson() ?: ""
}
```
And the same for variations.

JorgeMucientes replied:
I had second thoughts about that first. Since `meta_data` fields are not only about subscription details. So "encapsulating/hiding" it as an extension function inside `SubscriptionDetails.kt` would make less obvious in the future, where metadata parsing is happening and where to add additional parsing. However, I think is a minor thing with the benefit of removing duplication. So, I'm up for the refactor 👍🏼 !

hichamboushaba reviewed (approved):
Great work @JorgeMucientes, it works well, I left one suggestion, but it's not a blocking point.

JorgeMucientes committed:
Extract extension function into Subscription details

JorgeMucientes committed:
Update fluxC changeset to latest trunk commit

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/model/SubscriptionDetails.kt:
```diff
@@ -15,3 +19,28 @@ data class SubscriptionDetails(
     val oneTimeShipping: Boolean
 ) : Parcelable
+
+// Currently, only subscription details metadata is editable from the app. The rest is read only and thus
+// not included in the persisted metadata.
```
np, @JorgeMucientes this comment is out of place now, can we move it to a function in `Product` itself?

To make this easier, we can probably make the function `toMetadataJson` return the raw `JsonArray` itself, and handle the conversion to `String` in the function in `Product`/`Variation`, this way, if we need to add more metadata in the future, we'll be able to benefit from `JsonArray` helper functions `add` and `addAll`, this is a minor point, just a suggestion 🙂

JorgeMucientes replied:
Good point about returning the JsonArray() for easier chaining in the future. Thanks for all the tips @hichamboushaba 👍🏼

hichamboushaba replied:
Thanks for the changes @JorgeMucientes, :shipit:

JorgeMucientes committed:
Merge branch 'trunk' into feature/10124-sync-subscription-data

JorgeMucientes committed:
Return json array instead of string

This will enable easier re-usability if in the future we need to parse additional data into json metadata

JorgeMucientes committed:
Update comment
